### PR TITLE
fix: Add swipe to dismiss for React Native Toast

### PR DESCRIPTION
**Pull request by amandaye0h**
## **Description**

Adds swipe-to-dismiss to the React Native `Toaster` (top-anchored Toast), porting the same behavior from metamask-mobile.

1. **What is the reason for the change?**
   - Toast covers navigation elements when it appears from the top. This prevents users from accessing back buttons.
   - Mobile already supports swipe-to-dismiss on legacy Toast and BaseNotification. The design-system `Toaster` should match so consumers get the same interaction without reimplementing it.
2. **What is the improvement/solution?** Users can swipe a toast upward to dismiss it (by distance or quick flick). Incomplete swipes spring back and resume the auto-dismiss timer. Auto-dismiss uses an interruptible `setTimeout` instead of Reanimated `withDelay`, so pans and toast replacement can pause/clear the timer correctly. Manual close and action handlers are unchanged.

Gesture thresholds match mobile (`0.35` height / `800` px/s / `-8` active offset Y / `20` fail offset X). Package adaptations: `scheduleOnRN` + explicit `'worklet'` directives for the prebuilt dist.

## **Related issues**

Related:
- https://github.com/MetaMask/metamask-mobile/pull/33598
- https://github.com/MetaMask/metamask-mobile/pull/33600

## **Manual testing steps**

1. Run `yarn storybook:ios` (or `yarn storybook:android`)
2. Open `Components/Toast` (or Toaster) and show a timed toast
3. Swipe the toast upward past ~35% of its height — it should animate out and dismiss
4. Show another toast and flick upward quickly — it should dismiss via velocity
5. Show a toast, swipe a short distance up, and release — it should spring back and still auto-dismiss on schedule
6. Show a persistent toast (`hasNoTimeout`) and swipe to dismiss — it should dismiss; taps/buttons should still work without accidental dismiss
7. Replace a timed toast with a persistent one — the new toast should stay visible (old timer must not fire)

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs)
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> **Medium Risk**
> Touches core toast lifecycle, timers, and gesture/Reanimated worklets; regressions could affect auto-dismiss, toast replacement, or dismiss while navigating, though behavior is heavily unit-tested.
> 
> **Overview**
> Adds **swipe-up to dismiss** on the design-system `Toaster`, aligned with metamask-mobile thresholds (distance, velocity, and pan activation offsets in `Toast.constants`).
> 
> `Toaster` now wraps the toast in a pan `GestureDetector`, tracks dismiss state and a **toast generation** so stale spring-back callbacks cannot resurrect a replaced toast, and routes manual close and swipe dismiss through shared dismiss animation. **Timed auto-dismiss** moves from Reanimated `withDelay` to an interruptible `setTimeout` that pauses on pan, resumes after spring-back (including elapsed-time logic), and is cleared when toasts are replaced so a persistent replacement is not dismissed by the previous timer. Layout-driven entrance runs once per toast via `animationStartedRef`.
> 
> Tests add gesture-handler mocks and coverage for swipe paths, replacement/timer edge cases, and stale resume; Jest coverage thresholds for `Toaster.tsx` are relaxed for `isDismissing` branches that are hard to assert under the Reanimated mock.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a0d9b90a43e2ec1f1af797c949dbd3981fa29a94. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/design-system-react-native/src/components/Toast/Toaster.tsx [hunk 1]
```diff
@@ -113,15 +207,112 @@ const ToasterComponent = forwardRef<ToasterRef, ToasterProps>(
...
+        Gesture.Pan()
+          .activeOffsetY(TOAST_SWIPE_ACTIVE_OFFSET_Y)
+          .failOffsetX([-TOAST_SWIPE_FAIL_OFFSET_X, TOAST_SWIPE_FAIL_OFFSET_X])
+          .onStart(() => {
+            'worklet';
+
+            // Don't interrupt an in-progress dismiss animation.
+            if (isDismissing.value) {
+              return;
+            }
```
Comment (georgewrmarshall) [inline packages/design-system-react-native/src/components/Toast/Toaster.tsx:253]: non-blocking: `failOffsetX` is configured, so a pan that activates on the Y axis but then exceeds horizontal travel will fail — `onEnd` is skipped and the toast can be left stuck mid-offset. BaseNotification (#33600) handles this with an `onFinalize` handler that calls `springBackAfterSwipe` when `isSwipeActive` is still true after a failed pan. Worth adding the same here.

```ts
.onFinalize(() => {
  'worklet';
  if (isDismissing.value) return;
  // spring back if onEnd was skipped (e.g. failOffsetX triggered)
  translateYProgress.value = withSpring(visibleTranslateY.value, TOAST_SPRING_CONFIG, ...);
})
```

File: packages/design-system-react-native/src/components/Toast/Toaster.tsx [hunk 2]
```diff
@@ -113,15 +207,112 @@ const ToasterComponent = forwardRef<ToasterRef, ToasterProps>(
...
+              TOAST_SPRING_CONFIG,
+              (finished) => {
+                // A new pan cancels this spring via cancelAnimation; only resume
+                // auto-dismiss when the spring-back completed naturally.
+                if (finished) {
+                  scheduleOnRN(
+                    resumeAutoDismissFromSwipe,
+                    generationAtSpringStart,
+                  );
+                }
```
Comment (georgewrmarshall) [inline packages/design-system-react-native/src/components/Toast/Toaster.tsx:308]: non-blocking: The `24` minimum dismiss distance is a magic number. BaseNotification (#33600) already named this as `NOTIFICATION_DISMISS_MIN_DISTANCE = 24`. Worth doing the same here — e.g. `TOAST_DISMISS_MIN_DISTANCE` in `Toast.constants.ts` — so the floor is visible and adjustable alongside the other gesture thresholds.

**Review by georgewrmarshall — APPROVED** [review]
Clean port with a few genuine improvements over the legacy mobile PRs: `scheduleAutoDismiss` correctly uses `setTimeout` (consistent with Toast #33598, avoids the `withDelay` inconsistency in BaseNotification #33600), `resumeAutoDismissAfterSwipe` consolidates the two separate entrance-interrupted / normal-resume functions into one, and the generation counter is a cleaner stale-spring-back guard than the null-check approach in the mobile PRs. Two non-blocking notes below.
